service/storage_service.cc

```diff
@@ -1876,33 +1876,56 @@ future<std::map<gms::inet_address, float>> storage_service::effective_ownership(
-        for (auto endpoints : ss._token_metadata.get_topology().get_datacenter_endpoints()) {
-            // calculate the ownership with replication and add the endpoint to the final ownership map
-            for (const gms::inet_address& endpoint : endpoints.second) {
-                float ownership = 0.0f;
-                for (range<token> r : ss.get_ranges_for_endpoint(keyspace_name, endpoint)) {
-                    // get_ranges_for_endpoint will unwrap the first range.
-                    // With t0 t1 t2 t3, the first range (t3,t0] will be splitted
-                    // as (min,t0] and (t3,max]. Skippping the range (t3,max]
-                    // we will get the correct ownership number as if the first
-                    // range were not splitted.
-                    if (!r.end()) {
-                        continue;
-                    }
-                    auto end_token = r.end()->value();
-                    if (token_ownership.find(end_token) != token_ownership.end()) {
-                        ownership += token_ownership[end_token];
-                    }
-                }
-                final_ownership[endpoint] = ownership;
-            }
-        }
-        return final_ownership;
+
+        auto final_ownership = make_lw_shared<std::map<gms::inet_address, float>>();
+        // The following loops seems computationally heavy, but it's not as bad.
+        // The upper two simply iterate over all the endpoints by iterating over all the
+        // DC and all the instances in each DC.
+        //
+        // The call for get_range_for_endpoint is done once per endpoint
+        return do_with(std::map<token, float>(dht::token::describe_ownership(ss._token_metadata.sorted_tokens())),
+                std::unordered_map<sstring,std::unordered_set<gms::inet_address>>(ss._token_metadata.get_topology().get_datacenter_endpoints()),
+                [&ss, keyspace_name, final_ownership](const std::map<token, float>& token_ownership, std::unordered_map<sstring,
+                        std::unordered_set<gms::inet_address>>& datacenter_endpoints) {
+            return do_for_each(datacenter_endpoints, [&ss, &keyspace_name, final_ownership, &token_ownership, &datacenter_endpoints](std::pair<sstring,std::unordered_set<inet_address>>&& endpoints) {
+                return do_with(std::unordered_set<inet_address>(endpoints.second), [&ss, &keyspace_name, final_ownership, &token_ownership, &endpoints](const std::unordered_set<inet_address>& endpoints_map){
+                    return do_for_each(endpoints_map, [&ss, &keyspace_name, final_ownership, &token_ownership, &endpoints](const gms::inet_address& endpoint) {
+                        // calculate the ownership with replication and add the endpoint to the final ownership map
+                        try {
+                            return do_with(float(0.0f), dht::token_range_vector(ss.get_ranges_for_endpoint(keyspace_name, endpoint)),
```

Comment:
get_ranges_for_endpoint() is also a nested loop!

Reply:
it is, but not a too bad one, assuming that we are preempting between each call, I belive it should be ok